getExcelAPA

grewered opened the pull request:
I added the new function(s). write now there are 3, I could merge them into one, but I thought this might be better for the overview. 

I also added the packages `tidyxl` for the formatting info and `checkmate`. And I added myself as an contributor on the package, I'm not sure with what role though. I'm not sure what the procedure is in this situation.

Commit:
added new function `getExcelAPA`

Commit:
added new packages and author to DESCRIPTION

beckerbenj commented on R/getExcelAPA.R:
```diff
@@ -0,0 +1,77 @@
+
+addItalic <- function(filePath, sheet){
+  # reading in Excel with format
+  format <- xlsx_cells(filePath, sheet)
```
`tidyxl::xlsx_cells()`

beckerbenj commented on R/getExcelAPA.R:
```diff
@@ -0,0 +1,77 @@
+  return(excelTable)
+}
+
+addItalic <- function(filePath, sheet){
```
I like that you split the function up into `getExcelAPA()`, `addItalic()` and `addURL()`. Would it be possible to split the functions in a way that `addItalic()` and `addURL()` both receive only a character vector to encapsule this specific functionality? Maybe importing from `.xlsx` could live in the main function or in a separate sub-function?

grewered replied:
I will look at that. But it might be difficult because when I just give them character vectors, the formatting information is lost (italic). with `addURL` I can see. For now they both receive data frames (no filePaths).

beckerbenj reviewed (changes requested):
Thanks for adding this function to `eatCodebook`! I think this provides great functionaltiy and could save substantial manual work (which is also error prone).

Could you please add some tests for these functions? This would also make reviewing the actual string substitution parts easier for me.

In the inline comments you can find some more specific remarks.

Commit:
updated DESCRIPTION

Commit:
updated sheet and df checks/selection. added new check function `check_litInfoAPA`. reading in file now only in main function `getExcelAPA`

Commit:
renamed functions. main function now called `getAPAInfo`, check function `check_APAInfo`

beckerbenj commented on DESCRIPTION:
```diff
@@ -5,8 +5,8 @@ Authors@R: c(
     person("Felix", "Milles", role = c("aut")),
-    person("Johanna", "Busse", role = c("aut")),
-    person("Edna", "Grewers", role = c("aut")))
+    person("Johanna", "Busse", role = c("ctb")),
```
LGTM

beckerbenj commented on R/getExcelAPA.R:
```diff
@@ -7,36 +7,47 @@
+    excelTable_format <- tidyxl::xlsx_cells(filePath, include_blank_cells = FALSE)
+  } else {
+    for(i in 1:length(excelTable)){
+      if(is.null(check_APAInfo(excelTable[[i]]))){
```
Could this be simplified into:

```
check_APAInfo(excelTable[[i]])
excelTable_format <- tidyxl::xlsx_cells(filePath, include_blank_cells = FALSE,
                                                sheets = i)
```

grewered replied:
line 28: doesn't work. instead: new function `is_APAInfo()` (output TRUE or FALSE). and line changed to `if(is_APAInfo(excelTable[[i]])){`

> is_APAInfo <- function(excelTable) {
>   #browser()
>  if(!is.data.frame(excelTable)) return(FALSE)
>   if(!identical(names(excelTable), c("Kurzangabe", "Langangabe"))) return(FALSE)
>   return(TRUE)
> }

beckerbenj commented on R/getExcelAPA.R:
```diff
@@ -7,36 +7,47 @@
+      if(is.null(check_APAInfo(excelTable[[i]]))){
+        excelTable_format <- tidyxl::xlsx_cells(filePath, include_blank_cells = FALSE,
+                                                sheets = i)
+        excelTable <- excelTable[[i]]
```
I am very unsure about this line. What happens here? Are you not looping over `excelTable`? What happens if `excelTable` has `length > 1`?

beckerbenj commented on R/getExcelAPA.R:
```diff
@@ -7,36 +7,47 @@
+
   # adding latex syntax for italic and URL input
-  long_new <- addItalic(filePath, sheet)
+  long_new <- addItalic(format)
```
LGTM!

beckerbenj commented on R/getExcelAPA.R:
```diff
@@ -75,3 +86,13 @@ addURL <- function(lang_neu){
 }
 
+
+check_APAInfo <- function(excelTable) {
```
LGTM!

beckerbenj reviewed (changes requested):
Thanks for adding these changes! I am very sorry for the late code review.

Thank you for the clarification regarding the scope/purpose of the function! That makes sense. Could some of this reasoning be added to the function documentation?

Also adding tests would it make substantially easier for me to review these functions. Would love those!

Disclaimer:
I am well aware that `eatCodebook` is a tricky place when it comes to adding code, because the existing code is very messy, unorganized, is not well tested and often buggy. 

We can also discuss these things during our meeting tomorrow!

grewered commented on R/getExcelAPA.R:
```diff
@@ -0,0 +1,77 @@
+#############################################################################
+#' Import APA references from an Excel sheet.
+#'
+#' Import APA references from one sheet from an Excel file, while translating italic formatting and URLs into proper LaTeX code.
```
add sheet argument

Commit:
fixed typo

Commit:
updated `addLatex_italic` and worked on `addLatex_URL`

Commit:
updated `addURL_latex`, cleaned up function `getAPAInfo`

Commit:
added some tests for `getAPAInfo` and example Excel files

Commit:
fixed typo in addLatex_italic

Commit:
added and updated tests for `getAPAInfo()`

Commit:
added documentation for `getAPAInfo()`

Commit:
`getAPAInfo()`: added example in documentation and example Excel in `extdata`

grewered commented:
@beckerbenj I updated the function and the helper functions, wrote some tests and updated the documentation. could you review it, please? If you have the time. I made the following updates: 

- I renamed some arguments to hopefully be better to understand. 
- I rewrote the helper functions, so that they both just get a character vector to which they add latex syntax
- check_APAInfo: returns NULL or an error message
- is_APAInfo: returns TRUE or FALSE
- I added an example in the documentation

I wasn't sure how to look at the documentation in the "help" menu while working on it. And I wasn't sure what to add to `@export` in the documentation, if I need to add anything there.

Commit:
updated example in documentation

beckerbenj reviewed (changes requested):
The function looks already a lot better to me! I especially like the cleanness of the helper functions.

My remaining main concern is regarding the list/data.frame input. Would it be sufficient to narrow down the allowed input format to a specific one?

Commit:
fixed typo: urlstyle

grewered commented on R/getAPAInfo.R:
```diff
@@ -0,0 +1,112 @@
+  if(is_APAInfo(ref_table)){
+    ref_table_format <- tidyxl::xlsx_cells(filePath, include_blank_cells = FALSE)
+  } else {
+    for(i in 1:length(ref_table)){
```
should I add why in a comment? it's because the BT team usually has reference Excel files with two sheets: one where they match the in-text citations to the variables, and one with a list of in-text citations and proper references

Commit:
updated documentation and added commentary in code

Commit:
added a sheet argument and added/updated the documentation/man file

Commit:
worked on new tests and updated the function.

Commit:
finished updating the tests. all pass

grewered commented:
I added a sheet argument and updated the tests. can you review?

beckerbenj commented on R/getAPAInfo.R:
```diff
@@ -2,41 +2,45 @@
-        ref_table <- ref_table[[i]]
-      }
+
+  if(sheet == 1){
```
Could this be simplified, in the sense of: Is there a need for a separate if condition `sheet == 1`? For example:

```
ref_table <- getExcel(filePath)
if(!is_APAInfo(ref_table)){
      ref_table <- ref_table[[sheet]]
}
```

beckerbenj commented on tests/testthat/test-getAPAInfo.R:
```diff
@@ -18,20 +18,46 @@ test_that("Import Excel with two sheets, select proper sheet", {
 
 test_that("Import Excel without reference sheet", {
-  expect_error(references <- getAPAInfo("helper_getAPAInfo_error1.xlsx"))
+  expect_error(references <- getAPAInfo("helper_getAPAInfo_error1.xlsx", sheet = 1))
```
Could you add the specific error message here in the tests?

beckerbenj commented on tests/testthat/test-getAPAInfo.R:
```diff
@@ -18,20 +18,46 @@ test_that("Import Excel with two sheets, select proper sheet", {
+
+test_that("wrong sheet selected", {
+  # file with only one sheet
+  expect_error(references <- getAPAInfo("helper_getAPAInfo_1sheet.xlsx"))
```
See above

beckerbenj reviewed (changes requested):
All in all looks very good to me!

Feel free to merge after incorporating the few inline comments :+1:

Commit:
updated tests and input validation

grewered commented:
@beckerbenj 
there's some issue with package dependencies. Is that something that I can/should change or is it a general eatCodebook issue?
Otherwise I would merge.

beckerbenj commented:
> Is that something that I can/should change or is it a general eatCodebook issue?

That is a general issue; feel free to merge!
